bundle/edu.gemini.pot/src/main/scala/edu/gemini/spModel/target/TargetParamSetCodecs.scala

```diff
@@ -0,0 +1,124 @@
+    ParamSetCodec.initial(SiderealTarget.empty)
+      .withParam("name",                     SiderealTarget.name)
+      .withOptionalParam("radial-velocity",  SiderealTarget.radialVelocity)
+      .withOptionalParam("redshift",         SiderealTarget.redshift)
+      .withOptionalParam("parallax",         SiderealTarget.parallax)
+      .withManyParamSet("magnitude",         SiderealTarget.magnitudes)
+      .withParamSet("coordinates",           SiderealTarget.coordinates)
+      .withOptionalParamSet("proper-motion", SiderealTarget.properMotion)
+
+  implicit val TooTargetParamSetCodec: ParamSetCodec[TooTarget] =
+    ParamSetCodec.initial(TooTarget.empty)
+      .withParam("name", TooTarget.name)
+
+  implicit val CometParamSetCodec: ParamSetCodec[HorizonsDesignation.Comet] =
+    ParamSetCodec.initial(HorizonsDesignation.Comet(""))
+      .withParam("des", HorizonsDesignation.Comet.des)
+      
+  implicit val AsteroidParamSetCodec: ParamSetCodec[HorizonsDesignation.Asteroid] =
+    ParamSetCodec.initial(HorizonsDesignation.Asteroid(""))
+      .withParam("des", HorizonsDesignation.Asteroid.des)
+      
+  implicit val AsteroidOldStyleParamSetCodec: ParamSetCodec[HorizonsDesignation.AsteroidOldStyle] =
+    ParamSetCodec.initial(HorizonsDesignation.AsteroidOldStyle(0))
+      .withParam("num", HorizonsDesignation.AsteroidOldStyle.num)
+      
+  implicit val MajorBodyParamSetCodec: ParamSetCodec[HorizonsDesignation.MajorBody] =
+    ParamSetCodec.initial(HorizonsDesignation.MajorBody(0))
+      .withParam("num", HorizonsDesignation.MajorBody.num)
+      
+  implicit val HorizonsDesignationParamSetCodec: ParamSetCodec[HorizonsDesignation] =
+    new ParamSetCodec[HorizonsDesignation] {
+      val pf = new PioXmlFactory
+      def encode(key: String, a: HorizonsDesignation): ParamSet = {
+        val (tag, ps) = a match {
+          case d: HorizonsDesignation.Comet            => ("comet", CometParamSetCodec.encode(key, d))
+          case d: HorizonsDesignation.Asteroid         => ("asteroid", AsteroidParamSetCodec.encode(key, d))
+          case d: HorizonsDesignation.AsteroidOldStyle => ("asteroid-old-style", AsteroidOldStyleParamSetCodec.encode(key, d))
+          case d: HorizonsDesignation.MajorBody        => ("major-body", MajorBodyParamSetCodec.encode(key, d))
+        }
+        Pio.addParam(pf, ps, "tag", tag)
+        ps
+      }    
+      def decode(ps: ParamSet): PioError \/ HorizonsDesignation = {
+        (Option(ps.getParam("tag")).map(_.getValue) \/> MissingKey("tag")) flatMap {
+          case "comet"              => CometParamSetCodec.decode(ps)
+          case "asteroid"           => AsteroidParamSetCodec.decode(ps)
+          case "asteroid-old-style" => AsteroidOldStyleParamSetCodec.decode(ps)
+          case "major-body"         => MajorBodyParamSetCodec.decode(ps)
+          case hmm                  => UnknownTag(hmm, "HorizonsDesignation").left
+        }
+      }
+    }
```

Comment:
The mechanism for doing tagged sum types is ugly, and needs to be factored out, at least a little.